crates/rpc_client/src/lib.rs

```diff
@@ -0,0 +1,650 @@
+//    const rpcBlock = await this._jsonRpcClient.getBlockByNumber(
+//      const noncePromise = this._jsonRpcClient.getTransactionCount(
+//      const accountData = await this._jsonRpcClient.getAccountData(
+//    const remoteValue = await this._jsonRpcClient.getStorageAt(
+
+#[derive(thiserror::Error, Debug)]
+pub enum RpcClientError {
+    #[error("Response was not of the expected type")]
+    InterpretationError {
+        msg: String,
+        request_body: String,
+        expected_type: String,
+        response_text: String,
+    },
+
+    #[error("Failed to send request")]
+    SendError { msg: String, request_body: String },
+
+    #[error("Failed to get response body")]
+    ResponseError { msg: String, request_body: String },
+
+    #[error(transparent)]
+    OtherError(#[from] std::io::Error),
+}
+
+pub struct AccountData {
+    pub balance: U256,
+    pub code: Bytes,
+    pub transaction_count: U256,
+}
+
+pub struct RpcClient {
+    url: String,
+    client: reqwest::Client,
+    next_id: AtomicU64,
+}
+
+struct Request {
+    body: String,
+    id: jsonrpc::Id,
+}
+
+impl RpcClient {
+    pub fn new(url: &str) -> Self {
+        RpcClient {
+            url: url.to_string(),
+            client: reqwest::Client::new(),
+            next_id: AtomicU64::new(0),
+        }
+    }
+
+    fn make_request(&self, method: &str, params: String) -> Request {
+        let id = jsonrpc::Id::Num(self.next_id.fetch_add(1, Ordering::Relaxed));
+
+        Request {
+            body: format!(
+                "{{
+                    \"jsonrpc\":\"2.0\",
+                    \"method\":\"{}\",
+                    \"params\":[{}],
+                    \"id\":{}
+                }}",
+                method,
+                params,
+                serde_json::json!(id)
+            ),
+            id,
+        }
+    }
+
+    async fn submit_request(&self, request_body: &str) -> Result<String, RpcClientError> {
+        use RpcClientError::{ResponseError, SendError};
+
+        self.client
+            .post(self.url.to_string())
+            .body(request_body.to_string())
+            .send()
+            .await
+            .map_err(|err| SendError {
+                msg: err.to_string(),
+                request_body: request_body.to_string(),
+            })?
+            .text()
+            .await
+            .map_err(|err| ResponseError {
+                msg: err.to_string(),
+                request_body: request_body.to_string(),
+            })
+    }
+
+    pub async fn get_tx_by_hash(&self, tx_hash: &B256) -> Result<eth::Transaction, RpcClientError> {
+        let request = self.make_request(
+            "eth_getTransactionByHash",
+            format!("{}", serde_json::json!(tx_hash)),
+        );
+
+        let response_text = self.submit_request(&request.body).await?;
+
+        let success: jsonrpc::Success<eth::Transaction> = serde_json::from_str(&response_text)
```

Comment:
if this is important, i could use some guidance on it. i tried extracting a method to de-duplicate the success handling:

```Rust
    fn verify_success<T>(response: Response, expected_type: &str) -> Result<T, RpcClientError>
    {
        let success: jsonrpc::Success<T> = serde_json::from_str(&response.text)
            .map_err(|err| RpcClientError::InterpretationError {
                msg: err.to_string(),
                request_body: response.request_body,
                expected_type: String::from(expected_type),
                response_text: response.text,
            })?;

        assert_eq!(success.id, response.request_id);

        Ok(success.result)
    }
```

But it wouldn't compile:
![image](https://user-images.githubusercontent.com/7883777/211480869-0c7ebbb6-162c-4b5b-a98b-aac644ab4d6f.png)

I tried to follow the help (`consider restricting type parameter T`), so I added a `where` clause, but that didn't work because some (all?) of the types I want to restrict by are structs, not traits:

![image](https://user-images.githubusercontent.com/7883777/211481284-af26bde7-96f4-49b2-a74f-823c027e2c88.png)

what am i missing here? please enlighten me.

Reply:
We can discuss this during today's call.

Reply:
I tried to take a stab at this, but, after spending a decent amount of time on it, the solution still wasn't clear to me, so i decided to spend my time on getting working code in place elsewhere. if you'd be inclined to make the change yourself and push the commit up here, i'd be happy to learn from your example.